code/obj/item/gun/energy.dm

```diff
@@ -214,26 +216,185 @@ TYPEINFO(/obj/item/gun/energy)
+	desc = "A showpiece model of the original Hafgan Mod.00 prototype laser pistol. "
 	muzzle_flash = "muzzle_flash_laser"
-	cell_type = null
+	cell_type = /obj/item/ammo/power_cell/fake
 	uses_charge_overlay = TRUE
 	charge_icon_state = "caplaser"
+	var/panelOpen = FALSE //open it with a screwdriver to modify it
+	var/obj/item/lens/myLens = null
+	var/obj/item/coil/small/myCoil = null
+	var/customProjectileDamage = CAPTAINGUN_T1_DAMAGE
+	var/customProjectileCost = CAPTAINGUN_T1_COST
 
 	New()
-		if (!src.current_projectile)
-			src.set_current_projectile(new /datum/projectile/laser/glitter)
-		if (isnull(src.projectiles))
-			src.projectiles = list(src.current_projectile)
-		..()
-		src.UpdateIcon()
+		. = ..()
+		src.set_current_projectile(null)
+		src.projectiles = null
+
+	examine()
+		. = ..()
+		if(src.evaluate_quality())
+			. += "It has had some parts replaced and is now a functional weapon. "
+		else
+			. += "While currently non-functional, it could be brought to working order with some replacement parts. "
+		if(src.panelOpen)
+			. += "Its maintenance panel is open. "
+			if(src.myLens)
+				. += "It contains a [src.myLens]. "
+			else
+				. += "It is missing a lens. "
+			if(src.myCoil)
+				. += "It contains a [src.myCoil]. "
+			else
+				. += "It is missing a small coil. "
+
+	attackby(obj/item/object, mob/user)
+		. = ..()
+		if(isscrewingtool(object))
+			user.show_text(SPAN_NOTICE("You [src.panelOpen ? "close" : "open"] the maintenance panel."))
+			playsound(user, 'sound/items/Screwdriver2.ogg', 65, TRUE)
+			src.panelOpen = !src.panelOpen
+			if(!src.panelOpen && src.evaluate_quality())
+				if(src.customProjectileDamage == CAPTAINGUN_T3_DAMAGE && src.customProjectileCost == CAPTAINGUN_T3_COST)
+					user.unlock_medal("Tinkerer", 1)
+		if(istype(object, /obj/item/coil/small))
+			if(src.panelOpen)
+				user.show_text(SPAN_NOTICE("You insert the [object] into the [src]."))
+				playsound(user, 'sound/items/Deconstruct.ogg', 65, TRUE)
+				user.put_in_hand_or_eject(src.myCoil)
+				object.set_loc(src)
+				user.u_equip(object)
+				src.myCoil = object
+			else
+				user.show_text(SPAN_NOTICE("The [src]'s maintenance panel is closed."))
+		if(istype(object, /obj/item/lens))
+			if(src.panelOpen)
+				user.show_text(SPAN_NOTICE("You insert the [object] into the [src]."))
+				playsound(user, 'sound/items/Deconstruct.ogg', 65, TRUE)
+				user.put_in_hand_or_eject(src.myLens)
+				object.set_loc(src)
+				user.u_equip(object)
+				src.myLens = object
+			else
+				user.show_text(SPAN_NOTICE("The [src]'s maintenance panel is closed."))
+
+	canshoot(mob/user)
+		if(!src.evaluate_quality())
+			return FALSE
+		//change the projectile here because this is checked by both gun.shoot_point_blank(), gun.shoot(), and gun.suicide()
+		//which duplicate a horrifying amount of code
+		src.set_current_projectile(new /datum/projectile/laser/custom(src.customProjectileDamage, src.customProjectileCost))
```

Comment:
God no, do not put side effects in `canshoot`. Can't you just set the projectile when the parts are updated?